### PR TITLE
[rush] unassigned operations can ignore weighting constraints

### PR DESCRIPTION
**Pull request by aramissennyeydd**
## Summary

We're dealing with some build cache inconsistencies across cobuild agents. We're seeing agents using the same lock, but unable to restore completed state as they don't have the same build cache ID. This bug is allowing 2 expensive operations to run side-by-side causing memory issues and timeouts related to memory pressure.

This PR uses the remote executing operation when reasonable and moves the unassigned status to just denote a sleep. It also improves the remote executing evaluation, since we no longer greedily get an event during execution time, we need to move that complexity into the scheduler. I added `checkAfter` and `lastCheckedAt` with the idea that someone could write a plugin to better set those values based on previous execution times, since waiting the default 5 seconds is wasteful.

## Details

From what I can tell, the unassigned operation needed to have a weight matching the possible operation it will pick up as https://github.com/microsoft/rushstack/blob/c1effc398416068b7f276bac32044f775643295d/libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts#L265 can start an operation that it reasons was not completed on another agent that locked it initially in the CacheableOperationPlugin, https://github.com/microsoft/rushstack/blob/c1effc398416068b7f276bac32044f775643295d/libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts#L388-L398. If the weight is not reflected onto the unassigned operation, multiple operations can be picked up and will skip the weight check. For example, say you have 3 operations with parallelism set to 8 and 2 machines:
1. Operation A with weight 8
2. Operation B with weight 8
3. Operation C with weight 4

Start: Machine 1 picks up Operation A, Machine 2 picks up Operation B
Step 1: Machine 1 finishes but fails to mark complete Operation A, Machine 2 finishes Operation B
Step 2: Machine 2 picks up Operation A and Operation C

The finishes but fails to mark complete is possible with a build cache ID inconsistency (what we're seeing) or if a machine gets lost during execution and doesn't report its success state, so another machine picks up the operation.

Removing the ability to assign a remote executing operation to a unassigned operation allows the weight to be correctly determined based on the specific operation that is being picked up.

## How it was tested

I tested this with the sharded repo and made sure that both cobuilds aren't deadlocking and that sharding is still working and working effectively. 

\nteresting sidenote, this change may open up the ability to do more dynamic wait times, which could help improve overall run times for agents that spend a lot of time in the sleep loop. Adjusting the sleep from 5s -> 1s dropped times from 40s -> 28s.

## Impacted documentation

None.

### CODE REVIEW

**Comment by chengcyber** [comment]
If I understand it correctly, the current challenge is to consistently reproduce the same schedule across the different cobuild processes. In the current lightweight implementation, there is no centralized scheduler to control which runner will be assigned to run a task.

To tackle with this, an initiative thought comes into my mind are:

1. Introduce a centralized scheduler
The centralized scheduler introduces a much more complicated and powerful architecture to distributed execution. The drawback is that it goes a little bit against the designated lightweight approach for cobuild.


2. Each runner can record how the tasks get scheduled, and cobuilds can reproduce the scheduler based on a log-based approach.

The workflow for 2 could be:
a: Turn on generating schedule logs for every runner by specifying a CLI parameter or ENV.
Running `RUSH_COBUILD_SCHEDULE_LOG=1 rush cobuild --to <package_name>` writes a log file somewhere, let's say `common/temp/cobuild.schedule.log`

b: Reproduce the schedule log by specifying a CLI parameter or ENV.
Running `RUSH_COBUILD_REPRODUCE_BY_SCHEDULE_LOG=1 rush cobuild -to <package_name>`, Rush instead of scheduling by `AsyncOperationQueue`, it invokes a similar OperationQueue instance (maybe ReproducibleAsyncOperationQueue) to accurate assign the task based on the `common/temp/cobuild.schedule.log` says.

-----

Back to this PR, `weight` property is actually in the same direction for Option 2. But it has some drawbacks I can tell:

1. Manually managing on weight setup is sort of loose and error-prune. A better approach is generated the "weight"（the generic idea I mean） from the actual run, such as `schedule.log` I proposed here.
2. It's not easy to use. For a cobuild runs a lot of tasks, `weight` setup comes trivial to use.

**Comment by aramissennyeydd** [comment]
@chengcyber I posted about this on Zulip, I think my path forward is to get rid of the unassigned status and move the sleep to`CacheableOperationPlugin`, that way we no longer need to interact with the scheduling loop at all or assign weights. 

I don't think we need a central scheduler at this point, we're just experiencing some growing pains with cobuilds and trying to lessen the pain. I'm working on a plugin at the moment to detect cache entry drift which is the other issue we're seeing.

**Comment by iclanton** [comment]
@dmichon-msft, can you take a look at this?

File: libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts [hunk 1]
```diff
@@ -261,8 +261,7 @@ export class OperationExecutionManager {
          */
         if (operation.status === UNASSIGNED_OPERATION) {
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts:262]: This `if` check will now be unreachable code.
We do, however, need something to periodically ping `assignOperations` since the assignability of the operations queue can change suddenly; I forget if that exists already.

Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts:262]: Ah, that's what I was running into. Was seeing a deadlock ~50% of the time because an operation would finish but there would be no way to get an item to queue. I'll take a look at adding this tomorrow.

Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts]: Updated in [5d86108](https://github.com/microsoft/rushstack/pull/4821/commits/5d86108a1ca39a30a31a8917b9d6f8c6ef0ae131) and [ad619fe](https://github.com/microsoft/rushstack/pull/4821/commits/ad619fea15f21f368a5d80cce61f0f413ef76365)

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts [hunk 2]
```diff
@@ -163,10 +163,17 @@ export class AsyncOperationQueue
...
       // remote executing operation which is not ready to process.
-      if (queue.some((operation) => operation.status === OperationStatus.RemoteExecuting)) {
+      const remoteExecutingOperation: OperationExecutionRecord | undefined =
+        this.tryGetRemoteExecutingOperation();
+      if (remoteExecutingOperation) {
+        waitingIterators.shift()!({
+          value: remoteExecutingOperation,
+          done: false
+        });
+      } else {
```
[the remark this answers is not in the document]
Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts]: Removed in [5d86108](https://github.com/microsoft/rushstack/pull/4821/commits/5d86108a1ca39a30a31a8917b9d6f8c6ef0ae131)

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts [hunk 3]
```diff
@@ -177,11 +180,13 @@ export class AsyncOperationQueue
...
   public tryGetRemoteExecutingOperation(): OperationExecutionRecord | undefined {
     const { _queue: queue } = this;
+
     // cycle through the queue to find the next operation that is executed remotely
     for (let i: number = queue.length - 1; i >= 0; i--) {
       const operation: OperationExecutionRecord = queue[i];
 
-      if (operation.status === OperationStatus.RemoteExecuting) {
+      const currentTime: number = new Date().getTime();
+      if (operation.status === OperationStatus.RemoteExecuting && operation.checkAfter < currentTime) {
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts:189]: Can we make "Remotely executing but a candidate to run locally anyway" into its own status code? Then have the code that sets the "RemoteExecuting" status also own the timeout to flag the operation as available to execute locally if the queue is empty.

Then we can just inject a call to `assignOperations` inside of `onOperationStatusChanged` and not need to do uncomfortable polling inside of the queue itself.

**Review by dmichon-msft — COMMENTED** [review]
This will work and is cleaner than what we had before, even if I still don't like having a polling system inside of the queue.

I think ideally we would wire up a handler for `onOperationStatusChanged` that calls `assignOperations`, then add some new status to indicate that an operation is remotely executing but a candidate for local execution, and have the code that sets `OperationStatus.RemoteExecuting` also set a timeout that then switches the status to that new special status value,

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts (hunk 3 above)
Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts]: Updated! Added a new status to reflect that "remote executing but might be ready for a check" status. Tried to also clean up some of the code that was particular to the remote executing status.

File: libraries/rush-lib/src/logic/operations/OperationExecutionRecord.ts [hunk 4]
```diff
@@ -166,6 +166,17 @@ export class OperationExecutionRecord implements IOperationRunnerContext {
     return this._operationMetadataManager?.stateFile.state?.cobuildRunnerId;
   }
 
+  public get isTerminal(): boolean {
+    return TERMINAL_STATUSES.has(this.status);
+  }
+
+  public get isRemoteExecuting(): boolean {
+    return (
+      this.status === OperationStatus.RemoteExecuting ||
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/OperationExecutionRecord.ts:175]: Status change to `RemoteExecuting` doesn't make new work available on the queue. We only need to kick `assignOperations()` if the status change caused an operation to go from unavailable for execution -> available.

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts [hunk 5]
```diff
@@ -129,7 +115,10 @@ export class AsyncOperationQueue
         // This operation is not yet ready to be executed
         // next one plz :)
         continue;
-      } else if (record.status === OperationStatus.RemoteExecuting) {
+      } else if (
+        record.status === OperationStatus.RemoteExecuting ||
+        record.status === OperationStatus.RemoteExecutingPossiblyComplete
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts:120]: I confess I hadn't realized that this was the cobuild _completion_ detector; my read of the initial implementation of the returning a RemoteExecuting operation had been that the runner was coded to execute the operation locally if it waiting longer than a certain delay with nothing to do.

As a result, we can simplify the logic a lot by just returning operations with status `RemoteExecutingPossiblyComplete` as ordinary candidates for execution. We can completely remove the `tryGetRemoteExecutingOperation` function, since as soon as operations in stathe `RemoteExecutingPossiblyComplete` get handed to the runner, their status will potentially change back to `RemoteExecuting`.

Basically the flow will be:
Ready -> Queued -> RemoteExecuting -(wait 500)-> RemoteExecutingPossiblyComplete -> Queued -> RemoteExecuting or terminal state

Comment (dmichon-msft) [reply libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts:120]: Alternatively, we could just reset the operation to `Ready` instead of having the separate `RemoteExecutingPossiblyComplete` status. If we keep the separate status, line 125 should be updated to include it as one of the possible states for an operation being returned from the primary loop.

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts [hunk 6]
```diff
@@ -166,9 +155,11 @@ export class AsyncOperationQueue
...
-      if (queue.some((operation) => operation.status === OperationStatus.RemoteExecuting)) {
+      const remoteExecutingOperation: OperationExecutionRecord | undefined =
+        this.tryGetRemoteExecutingOperation();
+      if (remoteExecutingOperation) {
         waitingIterators.shift()!({
-          value: { weight: 1, status: UNASSIGNED_OPERATION },
+          value: remoteExecutingOperation,
           done: false
         });
       }
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts:165]: This code path can go away in favor of just returning operations in state `OperationStatus.RemoteExecutingPossiblyComplete` as normal operations in the primary loop. If we don't want them to bounce through the `Queued` status we can add a check and not override before returning.

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts [hunk 7]
```diff
@@ -177,11 +168,12 @@ export class AsyncOperationQueue
...
+
     // cycle through the queue to find the next operation that is executed remotely
     for (let i: number = queue.length - 1; i >= 0; i--) {
       const operation: OperationExecutionRecord = queue[i];
 
-      if (operation.status === OperationStatus.RemoteExecuting) {
+      if (operation.status === OperationStatus.RemoteExecutingPossiblyComplete) {
         return operation;
       }
     }
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts:179]: This entire function is no longer needed, since remote executing operations to be checked are part of the primary loop.

File: libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts [hunk 8]
```diff
@@ -103,7 +98,12 @@ export class OperationExecutionManager {
     this._beforeExecuteOperation = beforeExecuteOperation;
     this._afterExecuteOperation = afterExecuteOperation;
     this._beforeExecuteOperations = beforeExecuteOperations;
-    this._onOperationStatusChanged = onOperationStatusChanged;
+    this._onOperationStatusChanged = (record: OperationExecutionRecord) => {
+      if (record.isRemoteExecuting) {
+        this._executionQueue.assignOperations();
+      }
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/OperationExecutionManager.ts:104]: Strictly speaking, the two status values for which we should invoke `assignOperations()` are `Ready` and `RemoteExecutingPossiblyComplete` (i.e. the statuses that make an operation become a candidate for execution).

**Review by dmichon-msft — COMMENTED** [review]
Thanks for your work on this so far. Reviewing this a few times has caused me to finally understand a bit better what exactly cobuilds is doing in the execution graph, and I think with this next round we can finally get to a comfortably simple model of how remote executing operations are handled.

Essentially, when cobuilds are enabled, an Operation has the following stages:
1. Waiting (for dependencies)
2. Ready (when all dependencies finished)
3. Queued
4. Try acquire cobuild lock
  a. If successful, goto 5
  b. If unsuccessful, mark as RemoteExecuting, then after 500ms goto 2
5. Restore from cache or Execute
6. Mark terminal status, release consumers

By listening to changes to operation.status values in OperationExecutionManager, we can ensure that we call `assignOperations` when operation statuses change to `Ready`.

File: libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts (hunk 5 above)
Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts:120]: I tried this, but ended up getting sidetracked trying to remove `RemoteExecuting` as a whole concept. All `RemoteExecuting` seems to be doing is allowing the queue to not mark an operation as complete and allow it to be requeued. Ideally, that would be possible if you return _any_ non-terminal status. If I return `Waiting` or `Executing` from a `beforeExecuteOperation` step, I think I'd want my operation to be requeued, not for the process to throw an error (as it does now). I ended up running into a few more issues trying to untangle the `RemoteExecuting` state and abandoned it, but I'd be curious to get your feedback on that direction?

Basically, we'd remove `RemoteExecuting`, `CacheableOperationPlugin` would return `Waiting` or `Executing` if it was unable to get the lock. `CacheOperationPlugin` would also set a timeout to set that operation's status to `Ready` after 500ms for it to be rechecked. The `onOperationStatusChanged` would be as you suggestion, only when the change is from x -> `Ready` do we trigger a call to `assignOperations`.

Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/AsyncOperationQueue.ts]: Updated to the above model in [7b7e7f1](https://github.com/microsoft/rushstack/pull/4821/commits/7b7e7f170478e5b55b8af3a3b4bd622953f8771f). Tested it a bit locally and didn't run into any issues.

File: libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts [hunk 9]
```diff
@@ -429,16 +433,12 @@ export class CacheableOperationPlugin implements IPhasedCommandPlugin {
...
-            return;
-          }
-          default: {
-            break;
-          }
+        if (!record.isTerminal || record.status === OperationStatus.NoOp) {
+          return;
         }
 
+        this._itemsToCheck.delete(record);
```
Comment (dmichon-msft) [inline libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts:440]: This line seems to be the only reference to this property; delete?

**Review by dmichon-msft — APPROVED** [review]
That's a lot simpler. I like. Just one minor bit of cleanup left with an unused property.

File: libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts (hunk 9 above)
Comment (aramissennyeydd) [reply libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts]: Leftover from another attempt. Removed in [0a86ddc](https://github.com/microsoft/rushstack/pull/4821/commits/0a86ddc0e43f233564ad23ac92f4e3c78c1a8a3f)

File: common/changes/@microsoft/rush/sennyeya-unassigned-operation-weight_2024-06-28-14-34.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Fixes an issue with cobuilds where weighted operations could ignore the scheduler's weighting requirements.",
```
Comment (aramissennyeydd) [inline common/changes/@microsoft/rush/sennyeya-unassigned-operation-weight_2024-06-28-14-34.json:5]:
```suggestion
      "comment": "Adjusts how cobuilt operations are added and requeued to the operation graph. Removes the 'RemoteExecuting' status.",
```

**Review by dmichon-msft — APPROVED** [review]
LGTM
